Added docs for X11 key remapping

relaxbox opened the pull request:
Lite-XL does not seem to be aware of any keyboard re-mapping done via the usual tools (such as `xkbset` or the GNOME settings UI). I assume this is because Lite is built on SDL, which is getting its keyboard input via some mechanism that is bypassing the usual X11 drudgery?

Anywhoo, as a lifelong Caps -> Ctrl remapper this is kind of a big deal. Fortunately there's a pretty straightforward workaround. This is a big enough deal that I think it'll cause a lot of people to bounce off Lite-XL pretty quickly.

I'm not sure this is the best place in the docs but I think it's worth documenting *somewhere*. Happy to move locations or change wording as suggested.

Guldoman commented on locales/en/tutorials/simple-plugin.md:
```diff
@@ -163,12 +163,14 @@ command.add(nil, {
+            config.plugins.simple.hw_message = text
+            config.plugins.simple.show_my_message = true
+         end
+         })
```
Just a nit, I'd remove an indent level here.

Guldoman commented:
Awesome, thanks!

relaxbox committed:
Added docs for X11 key remapping

Lite-XL does not seem to be aware of any keyboard re-mapping done via the usual tools (such as `xkbset` or the GNOME setting).

I assume this is because Lite is built on SDL, which is getting its keyboard input via some mechanism that is bypassing the usual X11 drudgery?

Anywhoo, as a lifelong Caps -> Ctrl remapping this is kind of a big deal. Fortunately there's a pretty straightforward workaround. This is a big enough deal that I think it'll cause a lot of people to bounce off Lite-XL pretty quickly.

I'm not sure this is the best place in the docs but I think it's worth documenting *somewhere*. Happy to move locations or wording as suggested.

takase1121 commented on docs/user-guide/configuration.md:
```diff
@@ -236,6 +236,15 @@ To remove an existing keyboard shortcut, you can use `keymap.unbind()`.
     Afterwards, press the "Save" button to save the changes.
     ![Screenshot showing the keybind changer][13]{ loading=lazy }
 
+### X11: Remapping CapsLock and Ctrl
```
I prefer this to be an [admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/).

Specifically

```
!!! tip "Keyboard remapping on X11"
    Lorem ipsum dolor si amet...
```

takase1121 commented on docs/user-guide/configuration.md:
```diff
@@ -236,6 +236,15 @@ To remove an existing keyboard shortcut, you can use `keymap.unbind()`.
 
+### X11: Remapping CapsLock and Ctrl
+
+Lite-XL reads scancodes from SDL so it's not aware of any remapping done via X11 (e.g. via `xkbset` or GNOME settings).
```
Do note that we use `Lite XL` in documentation, not `Lite-XL`, `LiteXL` or `lite-xl` (unless specifically referring to the binary itself).

Instead of mentioning scancodes which no one outside of gamedev would understand, you can explain that Lite XL determines shortcut keys based on their _physical_ location on the keyboard, regardless of what the keyboard layout was set in the OS (e.g. DVORAK).

relaxbox replied:
ok, thanks! I'll make those changes. Also, after looking around more, this page probably looks like a better home? https://lite-xl.com/setup/platform-notes/ 

If you agree I'll move it there.

takase1121 commented on docs/user-guide/configuration.md:
```diff
@@ -236,6 +236,15 @@ To remove an existing keyboard shortcut, you can use `keymap.unbind()`.
+### X11: Remapping CapsLock and Ctrl
+
+Lite-XL reads scancodes from SDL so it's not aware of any remapping done via X11 (e.g. via `xkbset` or GNOME settings).
+The tool `setscancodes` can be used as a workaround. A common use case is to map CapsLock to an additional Ctrl:
```
I see where you are coming from, but I don't want to put this in the documentation, at least in this form. `setscancodes` modifies the _kernel_ scancode map, which means this will affect all other applications until next boot. This is overkill. Also, if the user have XKB remap configured, wouldn't this mess with that?

I will recommend just telling the user _why_ XKB remap doesn't work, and suggest the user to either **update their shortcuts** (yes I know this is tedious), then suggest the `setscancode` workaround.

relaxbox replied:
> Also, if the user have XKB remap configured, wouldn't this mess with that?

It would if they are differing mappings (obviously), but if they're redundant then no it does not mess with it.

> I will recommend just telling the user why XKB remap doesn't work, and suggest the user to either update their shortcuts (yes I know this is tedious)

Is that even possible here? I don't think Lite XL recognizes caps lock as a modifier.

relaxbox replied:
> setscancodes modifies the kernel scancode map, which means this will affect all other applications until next boot. This is overkill.

Well, that's sort of the idea here -- globally remap the key across all applications, which is effectively what `xmodmap` et al do.

Guldoman replied:
> Is that even possible here? I don't think Lite XL recognizes caps lock as a modifier.

We do!
A plugin could be made to automatically switch caps and ctrl. Shouldn't be too complex.

Guldoman replied:
> We do!

*Needs to be set as a recognized keymap modkey, but it's basically two lines to add it.

I guess we could add this to the note directly:
```lua
local modkeys = require("core.modkeys-" .. (PLATFORM == "Mac OS X" and "macos" or "generic"))

modkeys.map["capslock"] = "ctrl" -- <- this maps capslock to ctrl
modkeys.map["left ctrl"] = "capslock" -- <- this maps left ctrl to capslock
table.insert(modkeys.keys, "capslock") -- <- this enables capslock as modifier
```

Guldoman replied:
I'm curious, is this workaround still needed on Lite XL master?

relaxbox replied:
Ah cool! I'm happy to totally biff this then and maybe just add a note to the docs, in that case.

takase1121 replied:
I'm partial towards this solution. We can remove the setkeycodes one.

takase1121 commented on docs/user-guide/configuration.md:
````diff
@@ -236,6 +236,15 @@ To remove an existing keyboard shortcut, you can use `keymap.unbind()`.
+The tool `setscancodes` can be used as a workaround. A common use case is to map CapsLock to an additional Ctrl:
+
+```bash
+sudo setkeycodes 3a 29
````
```suggestion
# setkeycodes 3a 29
```

relaxbox committed:
Various PR feedback

* Moved everything inside a "tip" admonition
* Used the proper name "Lite XL"
* Suggest remapping via the Lite XL config

relaxbox committed:
Fix typo

relaxbox commented:
@takase1121 I think I've made all your suggested changes:

* Moved everything inside an admonition
* Refer to the editor as "Lite XL"
* Use Guldoman's recommended remapping method

Let me know if there's anything else I can do

takase1121 committed:
Rephrase a little bit
